libs/geometry/src/Transcoder.cpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <geometry/Transcoder.h>
+
+#include <math/half.h>
+
+using filament::math::half;
+
+namespace filament {
+namespace geometry {
+
+// The internal workhorse function of the Transcoder, which takes arbitrary input but always
+// produced packed floats. We expose a more readable interface than this to users, who often have
+// untyped blobs of interleaved data.
+template<typename SOURCE_TYPE, int NORMALIZATION_FACTOR>
+void convert(float* target, void const* source, size_t count, int numComponents,
```

Comment:
Can you check (maybe with Mathias' help) that those loops get nicely SIMDified by the compiler? They seem to be perfect candidates for this (or even for job system for large data sets; if you have millions of vertices for instance).

Reply:
The resulting ARM code is fairly tight when using -O3 with clang.  For example, the multiply gets eliminated when it is 1 (which is why I made it consexpr) and the `scvtf` instruction is used for the cast:

https://gcc.godbolt.org/z/jfzMbfW8M

Reply:
The output is even better if you make `numComponents` a template argument: https://gcc.godbolt.org/z/ecsdxv6hx. Realistically numComponents will mostly be 3, maybe 1 or 2, so it might be worth it.

Reply:
I had thought of doing that but it is debatable if it is worth it.  The outer loop in my version has fewer instructions, yes there's a branch in there but this is a RISC architecture.

Reply:
Oh I see, it got simdified.  Sorry about that, haven't woken up yet.  Yeah I'll go ahead and do this for 3 & 4 components.

Reply:
Added a commit for this.